src/kernels/layernorm_kernels.cu

```diff
@@ -197,4 +253,54 @@ void layer_norm(torch::Tensor& out,
   });
 }
 
-}  // namespace llm::kernel
+template <typename T>
+void invoke_layernorm_kernel(T* out,
+                             const T* input,
+                             const T* weight,
+                             const T* bias,
+                             const float epsilon,
+                             int m,
+                             int n) {
+  layer_norm_kernel<T><<<m, n>>>(out, input, weight, bias, epsilon, n);
+}
+
+template <>
+void invoke_layernorm_kernel<half2>(half2* out,
```

Comment:
sounds this template specializations are optional since they are covered by the general template. no?